nni/retiarii/nn/pytorch/api.py

```diff
@@ -312,6 +313,125 @@ def __getitem__(self, item):
+        meaning that the block will be repeated at least `min` times and at most `max` times.
+    """
+    def __init__(self,
+                 blocks: Union[Callable[[], nn.Module], List[Callable[[], nn.Module]], nn.Module, List[nn.Module]],
+                 depth: Union[int, Tuple[int, int]], label=None):
+        super().__init__()
+        self._label = label if label is not None else f'valuechoice_{uid()}'
+        self.min_depth = depth if isinstance(depth, int) else depth[0]
+        self.max_depth = depth if isinstance(depth, int) else depth[1]
+        assert self.max_depth >= self.min_depth > 0
+        if not isinstance(blocks, list):
+            if isinstance(blocks, nn.Module):
+                blocks = [blocks] + [copy.deepcopy(blocks) for _ in range(self.max_depth - 1)]
+            else:
+                blocks = [blocks for _ in range(self.max_depth)]
+        assert len(blocks) > 0
+        if not isinstance(blocks[0], nn.Module):
+            blocks = [b() for b in blocks]
+        assert self.max_depth <= len(blocks)
+        self.blocks = nn.ModuleList(blocks)
+
+    @property
+    def label(self):
+        return self._label
+
+    def forward(self, x):
+        for block in self.blocks:
+            x = block(x)
+        return x
+
+
+class Cell(nn.Module):
+    """
+    Cell structure [1]_ [2]_ that is popularly used in NAS literature.
+
+    A cell consists of multiple "nodes". Each node is a sum of multiple operators. Each operator is chosen from
+    ``op_candidates``, and takes one input from previous nodes and predecessors. Predecessor means the input of cell.
+    The output of cell is the concatenation of some of the nodes in the cell (currently all the nodes).
+
+    Parameters
+    ----------
+    op_candidates : function or list of module
+        A list of modules to choose from, or a function that returns a list of modules.
+    num_nodes : int
+        Number of nodes in the cell.
+    num_ops_per_node: int
+        Number of operators in each node. The output of each node is the sum of all operators in the node.
+    num_predecessors : int
+        Number of inputs of the cell. The input to forward should be a list of tensors.
+    merge_op : str
+        Currently only ``all`` is supported, which has slight difference with that described in reference.
+    label : str
+        Identifier of the cell. Cell sharing the same label will semantically share the same choice.
+
+    References
+    ----------
+    .. [1] Barret Zoph, Quoc V. Le, "Neural Architecture Search with Reinforcement Learning". https://arxiv.org/abs/1611.01578
+    .. [2] Barret Zoph, Vijay Vasudevan, Jonathon Shlens, Quoc V. Le,
+        "Learning Transferable Architectures for Scalable Image Recognition". https://arxiv.org/abs/1707.07012
+    """
+
+    # TODO:
+    # Support loose end concat (shape inference on the following cells)
+    # How to dynamically create convolution with stride as the first node
+
+    def __init__(self,
```

Comment:
seems it is better to put this Cell class into search space zoo/hub

Reply:
Let's discuss on teams.